Polymer functionality via wD-MPNN

MarcS18 opened the pull request:
## Description
This PR implements the weighted directed message passing neural network originally proposed in the [Coley Group's work](https://github.com/coleygroup/polymer-chemprop) for modelling polymers. The original work was written for Chemprop v1, this PR aims to port over the method to Chemprop v2. This is achieved by adding weighting to the underlying MPNN, for small molecules these weight vectors default to 1 (will also allow user defined weightings). The initial edge-based message is calculated and multiplied by the bond weight (E_w), before aggregation each node's (atom) features are multiplied by the corresponding atom weight (V_w) and finally the graph representation is multiplied by the degree of polymerisation. For polymers, the new polymer featurizer generates the atom and bond weights, as well as the degree of polymerisation, used for weighting the molecular graph from the input string described in the [original paper](https://pubs.rsc.org/en/content/articlelanding/2022/SC/D2SC02839E).

The polymer functionality is implemented in the CLI in the same manner as reactions.

## Example / Current workflow
No current polymer functionality.

## Bugfix / Desired workflow
An example notebook is provided.

## Questions
The only function that hasn't currently been added is the change to the aggregation mean as I am currently not sure how to go about implementing this with the current aggregation implementation,

Original:
```python
if self.aggregation == 'mean':
    mol_vec = mol_vec.sum(dim=0) / a_size
```

Adjusted:     
```python
if self.aggregation == 'mean':
     mol_vec = mol_vec.sum(dim=0) / w_atom_vec.sum(dim=0)  # if not --polymer, w_atom_vec.sum == a_size
``` 

## Relevant issues
#476 

## Checklist
- [ x ] linted with flake8?
- [ x ] (if appropriate) unit tests added?

KnathanM commented:
> The only function that hasn't currently been added is the change to the aggregation mean as I am currently not sure how to go about implementing this with the current aggregation implementation,

Thinking at a high level maybe we could do something like:
```
class MeanAggregation:
    def forward(self, H, batch, V_w):
        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
        dim_size = batch.max().int() + 1
        sum_node_embeddings = torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, index_torch, H, reduce="sum", include_self=False
        )
        sum_of_weights = torch.zeros(dim_size, V_w.shape[1], dtype=V_w.dtype, device=V_w.device).scatter_reduce_(
            self.dim, index_torch, V_w, reduce="sum", include_self=False
        )
        return sum_node_embeddings / sum_of_weights
```

KnathanM commented:
@MarcS18 Thank you for the effort you put in to porting this over. Can you recommend one or two people to review the PR? Is there anyone else familiar with the v1 implementation that could look review this before one of the current maintainers takes a look? With a PR this big, it will probably take us a while.

MarcS18 commented:
> 
> Thinking at a high level maybe we could do something like:
> 
> ```
> class MeanAggregation:
>     def forward(self, H, batch, V_w):
>         index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
>         dim_size = batch.max().int() + 1
>         sum_node_embeddings = torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
>             self.dim, index_torch, H, reduce="sum", include_self=False
>         )
>         sum_of_weights = torch.zeros(dim_size, V_w.shape[1], dtype=V_w.dtype, device=V_w.device).scatter_reduce_(
>             self.dim, index_torch, V_w, reduce="sum", include_self=False
>         )
>         return sum_node_embeddings / sum_of_weights
> ```

Thanks for this, it looks like we can get the same transformation with:

```
class MeanAggregation:
    def forward(self, H: Tensor, batch: Tensor, V_w: Tensor, *args) -> Tensor:
        dim_size = batch.max().int() + 1
        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
        sum_node_embeddings = torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, index_torch, H, reduce="sum", include_self=False
        )
        V_w = V_w.unsqueeze(1)
        index_weight = batch.unsqueeze(1).repeat(1, V_w.shape[1])
        sum_of_weights = torch.zeros(dim_size, V_w.shape[1], dtype=V_w.dtype, device=V_w.device).scatter_reduce_(
            self.dim, index_weight, V_w, reduce="sum", include_self=False
        )
        return torch.div(sum_node_embeddings, sum_of_weights)
```
@KnathanM Unfortunately, I personally don't know anyone else familiar enough with polymer ML to be able to review this. It might be worth trying to contact the original repo owners, otherwise I think DeepChem are doing something similar so that's also an option.

I went ahead and removed the large datafile from the PR as it was only used for the example notebook and a reference to the original data source is already provided within it.

KnathanM commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+
+    def _get_atoms(self, rwmol: Chem.Mol, atom_features_extra: np.ndarray | None):
+        """
+        generates atom features for all 'core' atoms, i.e not R groups as previously tagged. This is done here to ensure that atoms linked to R groups
+        # have the correct saturation
```
```suggestion
        generates atom features for all 'core' atoms, i.e not R groups as previously tagged. This is
        done here to ensure that atoms linked to R groups have the correct saturation.
```

KnathanM commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+
+    def _get_atoms(self, rwmol: Chem.Mol, atom_features_extra: np.ndarray | None):
+        """
+        generates atom features for all 'core' atoms, i.e not R groups as previously tagged. This is done here to ensure that atoms linked to R groups
```
"this is done here", as opposed to where? It makes sense to generate the atom features when the featurizer is called, so I am not sure what the comment about correct saturation is about.

KnathanM commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+        n_atoms = rwmol.GetNumAtoms()
+        if n_atoms == 0:
+            V = np.zeros((1, self.atom_fdim), dtype=np.single)
+            V_w = np.ones((1, self.atom_fdim), dtype=np.single).flatten()
```
Why specify a 2D shape with n_rows=1, `(1, self.atom_fdim)`, and then immediately flatten it to 1D? you could just do `V_w = np.ones((self.atom_fdim), dtype=np.single)` I think.

KnathanM commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+            V = np.hstack((V, atom_features_extra))
+        # Check to ensure that the number of atoms equals the length of the extra atom features
+        n_atoms = len(V)
+        if atom_features_extra is not None and len(atom_features_extra) != n_atoms:
+            raise ValueError(
+                "Input molecule must have same number of atoms as `len(atom_features_extra)`!"
+                f"got: {n_atoms} and {len(atom_features_extra)}, respectively"
+            )
```
This check should go before the concatenation or else the error will be triggered by that ...

KnathanM commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+                ],
+                dtype=np.single,
+            )
+            V_w = np.array(
+                [
+                    a.GetDoubleProp("w_frag")
+                    for a in rwmol.GetAtoms()
+                    if a.GetBoolProp("core") is True
+                ],
+                dtype=np.single,
+            ).flatten()
```
This is fine, but seems inefficient to write and read the weight from each atom. Possibly we could just get the number of core atoms in each fragment and do `V_w = [w for w, n_atoms in zip(frag_weights, n_atoms_in_frags) for _ in range(n_atoms)]`.

KnathanM commented on chemprop/utils/utils.py:
```diff
@@ -89,6 +89,60 @@ def make_mol(
+    indicies = [a.GetIdx() for a in rwmol.GetAtoms() if "*" in a.GetSmarts()]
+    while len(indicies) > 0:
+        rwmol.RemoveAtom(indicies[0])
+        indicies = [a.GetIdx() for a in rwmol.GetAtoms() if "*" in a.GetSmarts()]
```
Why not just pop the index from the list instead of remaking the list each time?

MarcS18 replied:
The issue with using `pop` in this case is that `RemoveAtom()` reindexes the atoms in the molecule so the previous indicies are invalid and have to be found again.

KnathanM reviewed (commented):
Hi @MarcS18, sorry to take a while to review this. And in fact, I only got through about half of it at this point. Your documentation and examples are very helpful. Please start addressing my comments if you have time. One general principle is that we prefer to not make changes to existing Chemprop classes (core code, so not the CLI or tests) if possible to avoid introducing any unexpected backwards incompatibilities. Our preferred approach is to just make new classes that inherit if possible or duplicate code to a new class if needed. (For example you can't inherit from NamedTuples like `MolGraph`, but can make a new `PolymerMolGraph` that has all the same attributes plus the new ones you need.)

Also relevant to polymers is some people are interested in using BigSMILES notation and gBigSMILES with Chemprop, so we may try to extend your code here in the future to enable that.

MarcS18 commented:
>  One general principle is that we prefer to not make changes to existing Chemprop classes (core code, so not the CLI or tests) if possible to avoid introducing any unexpected backwards incompatibilities. Our preferred approach is to just make new classes that inherit if possible or duplicate code to a new class if needed. (For example you can't inherit from NamedTuples like `MolGraph`, but can make a new `PolymerMolGraph` that has all the same attributes plus the new ones you need.)

I did initially plan to add a separate `PolymerMolGraph` but getting it working was more challenging. Now that there are the `MolAtomBondGraph` and `BatchCuikMolGraph` examples this should be easier to do. This should also resolve the issues I was having with the `MolAtomBondMPNN` and clean up the tests. So I agree, now its functional, its a good time to go back and refactor this part.

MarcS18 commented on chemprop/data/dataloader.py:
```diff
@@ -15,7 +16,11 @@
+    | MolAtomBondDataset
+    | PolymerDataset
+    | ReactionDataset
+    | MulticomponentDataset,
```
I think for the moment its ok, it would just be adding an additional import as some of the types are used for instance checks later. Currently it is also clearer which dataset types can be used per function, which is probably more beneficial for the user, although if more are added it could become bloated later down the line.

MarcS18 commented on chemprop/data/datapoints.py:
```diff
@@ -149,6 +149,66 @@ def from_smi(
+
+
+@dataclass
+class PolymerDatapoint(_DatapointMixin, _PolymerDatapointMixin):
```
I'm hesitant to use `class PolymerDatapoint(_PolymerDatapointMixin, MoleculeDatapoint): pass` as once the polymers are separated from the MolGraph into a PolymerGraph they will become incompatible so having a distinction will be important. Maybe we just move these parts to an additional ` _mixin` as it is also repeated for `LazyMoleculeDatapoint`.

MarcS18 commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+        frag_weights = polymer.fragment_weights
+        rules = polymer.edges
+        self.polymer_info, self.degree_of_poly = self.parse_polymer_rules(rules)
+        # We apply atom weights here rather than in make_polymer_mol to avoid key errors when caching the atom properties
+        mol = self.apply_atom_weights(mol, frag_weights)
```
Yep, this should be `MolGraph` instead of atom properties

MarcS18 commented:
Updated with the following changes:

- The MolGraph for Polymers has been decoupled from the default `MolGraph` to a new `WeightedMolGraph`.
- Tests reset to defaults as `V_w` and `E_w` no longer required after decoupling of the `MolGraph`.
- Batching is now done in a separate `BatchWeightedMolGraph` and `collate_polymer_batch`.
- The MPNN for Polymers has been decoupled from the base `MPNN` to a new `wMPNN`.
- `BondMessagePassing` for `wMPNN` is now separated into `WeightedBondMessagePassing`.
- Introduced a new Mixin for `V_f`, `E_f` and `V_d` for Datapoints.
- `PolymerMolGraphFeaturizer` now consistently uses the `Polymer` type as input.
- Cleanup of `PolymerMolGraphFeaturizer` and CLI implementation.
- Reworked examples and documentation to reflect the above changes.
- No longer works with Multicomponents. I will likely do this as a separate PR after the base functionality has been implemented. There are some lines included in this PR that will help enable this in the future.
- Introduced optional `weighted` keyword into `SimpleMoleculeMolGraphFeaturizer` to enable Multicomponent models with `wMPNN`.
- Not advised to use `AtomMessagePassing` with `SumAggregation` due to overfitting.
- Black on notebooks for formatting.

MarcS18 commented on chemprop/data/datasets.py:
```diff
@@ -547,6 +556,40 @@ def reset(self):
+    @property
+    def polymers(self) -> list[PolymerDatapoint]:
+        """the polymers associated with the dataset"""
+        return self.data
```
This has now been changed to return the type Polymer for input to the `PolymerMolGraphFeaturizer`.

MarcS18 commented on chemprop/cli/utils/parsing.py:
```diff
@@ -345,6 +400,22 @@ def make_datapoints(
+                    keep_h,
+                    add_h,
+                    ignore_stereo=ignore_stereo,
+                    reorder_atoms=reorder_atoms,
```
Changed to all positional args.

MarcS18 commented on chemprop/cli/utils/parsing.py:
```diff
@@ -345,6 +400,22 @@ def make_datapoints(
             for mol_idx, smis in enumerate(smiss)
         ]
 
+        n_mols = len(smiss)
```
Removed.

MarcS18 commented on chemprop/data/datapoints.py:
```diff
@@ -201,6 +201,66 @@ def from_smi(
+    fragment_weights: list = field(default_factory=list)
+    """the fragment weights associated with each monomer"""
+    edges: str = None
+    """the directed edges between monomers"""
```
Arguments have been reordered to ensure `PolymerDatapoint(mol, y)` works aswell. `kw_only=True` was added to the `_mixins` to enable better ordering.

MarcS18 commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+        num_frags = len(mols)
+        if len(fragment_weights) != num_frags:
+            raise ValueError(
+                f"The number of input monomers/fragments ({num_frags}) does not match the number of input weights ({len(fragment_weights)})"
```
This should now be cleaned up to be more consistent.

MarcS18 commented on chemprop/featurizers/molgraph/polymer.py:
```diff
@@ -0,0 +1,364 @@
+    ) -> MolGraph:
+        mol = polymer.mol
+        frag_weights = polymer.fragment_weights
+        rules = polymer.edges
```
These have been changed to `edge_rules` to be consistent.
